Document managing users, roles and accesslists through IaC (#36622)
gravitational/teleport: docs/pages/management/dynamic-resources/access-list.mdx

---
title: Creating Access Lists with IaC
description: Use Infrastructure-as-Code tooling to create Teleport AccessLists.
---

> Access Lists allow Teleport users to be granted long term access to resources

Reviewer: ```suggestion
Access Lists allow Teleport users to be granted long-term access to resources
```
Because it's a compound adjective

> managed within Teleport. With Access Lists, administrators and access list
> owners can regularly audit and control membership to specific roles and
> traits, which then tie easily back into Teleport's existing RBAC system.

Reviewer: ```suggestion
managed within Teleport. With Access Lists, administrators can regularly audit and control membership to specific roles and
traits, which then tie easily back into Teleport's existing RBAC system.
```
Seemed a bit awkward to repeat "access list" twice with different capitalization, and I'm not sure it's clear this point what the difference between an admin and an Access List owner is.

In this guide, we'll follow up on [the IaC users and roles guide]()
by allowing users with the `manager` role to grant the `support-engineer` role
to users meeting specific criteria.

Please note that Access Lists can be managed via IaC but Access List memberships
cannot. The goal of Access Lists is to decentralize granting and reviewing
access. By allowing managers to grant access within specific guidelines and
automatically enforcing review, users can request common access rights without
having to go through the centralized team managing the Teleport IaC.
This reduces the load on the centralized IaC/security team, ensures the access
reviewer is aware of the context, reduces the request resolution time, and
ensures access grants are periodically reviewed.

> ### Prerequisites

Reviewer: We should bump each header level down one. This should be an H2.

> To follow this guide, you must follow first [the basic users and roles IaC guide]().

Reviewer 2: Missing link.

We will reuse its users and roles for our Access List.

> ### Step 1/3 - Write manifests

Reviewer: ```suggestion
## Step 1/3. Write manifests
```

#### Write the privileged role manifest

We will create a new role `support-engineer` that grants access to production
servers. The `engineer` role from the previous guide was only granting access to
`dev` and `staging` servers.

<Tabs>
<TabItem label="tctl">

Create the following `privileged-role.yaml` file:

```yaml
kind: role
version: v7
metadata:
  name: support-engineer
spec:
  allow:
    logins: ['root', 'ubuntu', '{{internal.logins}}']
    node_labels:
      'env': ['production']
```

</TabItem>

> <TabItem label="operator">

Reviewer: ```suggestion
<TabItem label="Operator">
```

Create the following `privileged-role.yaml` file:

```yaml
apiVersion: resources.teleport.dev/v5
kind: TeleportRole
metadata:
  name: support-engineer
spec:
  allow:
    logins: [ 'root', 'ubuntu', '{{internal.logins}}' ]
    node_labels:
      'env': [ 'production' ]
```

<Admonition type="note">
The role version is v5 because the Teleport Kubernetes Operator doesn't support
yet roles v6 and v7.

> </Admonition>

Reviewer 2: We're adding versions support pretty soon so maybe just remove this note now so we don't have to remove it in a couple of weeks?

</TabItem>

> <TabItem label="terraform">

Reviewer: ```suggestion
<TabItem label="Terraform">
```

Create the following `privileged-role.tf` file:

```hcl
resource "teleport_role" "support-engineer" {
  version = "v7"
  metadata = {
    name = "support-engineer"
  }

  spec = {
    allow = {
      logins = ["root", "ubuntu", "{{internal.logins}}"]
      node_labels = {
        env = ["production"]
      }
    }
  }
}
```

</TabItem>
</Tabs>

> #### Write the AccessList manifest

Reviewer: ```suggestion
#### Write the Access List manifest
```

> In this step we'll create an AccessList that allows users with the `manager`

Reviewer: ```suggestion
In this step we'll create an Access List that allows users with the `manager`
```
To refer to the general term and not the Kubernetes CRD

role such as  `alice` to grant access to production to users with the `engineer`
role.

<Tabs>
<TabItem label="tctl">

Create the following `accesslist.yaml` file:

```yaml
version: v1
kind: access_list
metadata:
  name: support-engineers
spec:
  title: "Production access for support engineers"
  audit:
    recurrence:

>       frequency: 6months[user-and-role.mdx](user-and-role.mdx)

Reviewer 2: Looks like a typo.
```suggestion
      frequency: 6months
```

  description: "Use this Access List to grant access to production to your engineers enrolled in the support rotation."
  owners:
    - description: "manager of NA support team"
      name: alice
  ownership_requires:
    roles:
      - manager
  grants:
    roles:

>       - support-engineer[user-and-role.mdx](user-and-role.mdx)

Reviewer 2: ```suggestion
      - support-engineer
```

  membership_requires:
    roles:
      - engineer
```

</TabItem>

> <TabItem label="operator">

Reviewer: ```suggestion
<TabItem label="Operator">
```

Create the following `accesslist.yaml` file:

```yaml
apiVersion: resources.teleport.dev/v1
kind: TeleportAccessList
metadata:
  name: support-engineers
spec:
  title: "Production access for support engineers"
  description: "Use this Access List to grant access to production to your engineers enrolled in the support rotation."
  audit:
    recurrence:
      frequency: 6months
  owners:
    - description: "manager of NA support team"
      name: alice
  ownership_requires:
    roles:
      - manager
  grants:
    roles:
      - support-engineer
  membership_requires:
    roles:
      - engineer
```

</TabItem>

> <TabItem label="terraform">

Reviewer: ```suggestion
<TabItem label="Terraform">
```

Create the following `accesslist.tf` file:

```hcl
resource "teleport_access_list" "support-engineers" {
  header =  {
    version = "v1"
    metadata = {
      name = "support-engineers"
    }
  }

  spec = {
    title = "Production access for support engineers"
    description = "Use this Access List to grant access to production to your engineers enrolled in the support rotation."
    audit = {
      recurrence = {
        frequency = 6
      }
    }
    owners = [
      {
        description = "manager of NA support team"
        name = "alice"
      }
    ]
    ownership_requires = {
      roles = ["manager"]
    }
    grants = {
      roles = ["support-engineer"]
    }
    membership_requires = {
      roles = ["engineer"]
    }
  }
}
```

</TabItem>
</Tabs>

### Step 2/3 - Apply the manifests

<Tabs>
<TabItem label="tctl">

```code
$ tctl create -f privileged-role.yaml
role 'support-engineer' has been created

$ tctl create -f accesslist.yaml
Access list "support-engineers" has been created
```

<Admonition type="note">
The user resource depends on roles, you must create roles before users as a user

> with a non-existing role is invalid and might be rejected by Teleport.

Reviewer 2: ```suggestion
with a non-existing role is invalid and will be rejected by Teleport.
```

</Admonition>

</TabItem>
<TabItem label="operator">

Create the Kubernetes CRs with the following commands:

```code
$ kubectl apply -n "$OPERATOR_NAMESPACE" -f privileged-role.yaml
teleportrole.resources.teleport.dev/support-engineer created

$ kubectl apply -n "$OPERATOR_NAMESPACE" -f accesslist.yaml
teleportaccesslist.resources.teleport.dev/support-engineers
```

</TabItem>
<TabItem label="terraform">

```code
$ terraform plan
[...]
Plan: 2 to add, 0 to change, 0 to destroy.

$ terraform apply
teleport_access_list.support-engineers: Creating...
teleport_role.support-engineer: Creating...
teleport_role.support-engineer: Creation complete after 0s [id=support-engineer]
teleport_access_list.support-engineers: Creation complete after 0s [id=support-engineers]
```

</TabItem>
</Tabs>

### Step 3/3 - Log in as `alice` and grant access to `bob`

Now, you created an Access List allowing `alice` to grant the `support-engineer`
role to its engineers.

You can log in as alice and add `bob` to the `support-engineers` Access List.

<Tabs>
<TabItem label="web UI">

Login as `alice` in the web UI, open the management panel and select the

> "AccessLists" tab. Your AccessList should be displayed, open it, choose "Enroll

Reviewer 2: ```suggestion
"Access Lists" tab. Your Access List should be displayed, open it, choose "Enroll
```
